mintscript/src/checker/timelock.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::{
+    chain::{block::timestamp::BlockTimestamp, timelock::OutputTimeLock},
+    primitives::{BlockDistance, BlockHeight},
+};
+
+pub trait TimelockChecker<C> {
+    type Error: std::error::Error;
+
+    /// Check timelock
+    fn check_timelock(&mut self, ctx: &mut C, lock: &OutputTimeLock) -> Result<(), Self::Error>;
```

Comment:
Why does the context need to be mutable?

Reply:
Strictly speaking it does not. The verification can be (and currently is) implemented without mutability. However, I was thinking of the following possible use cases:

* You can accumulate timelocks / signatures in a vec inside the context instead of checking right away. Already done in tests for checking iteration order.
* There can be a context adaptor that caches some information which requires updating some internal data structures as we go.
* Signature batching: On encountering a signature, just update some internal accumulator. When we're done with traversing input / transaction / block / whatever, check the accumulator.

All of the above could be achieved with interior mutability, not sure if that's better.

Reply:
If anything is only for tests, then we definitely can use a RefCell/Mutex for that. 

About the other ones, this is actually a scary trade-off, and I don't know which is better in an absolute fashion. Better performance or better security. I would prefer the latter, and then worry about performance later. 

An example where this was done well (I hope so, at least) is in the transaction verifier. We isolated mutability from chainstate by making the base reference of TransactionVerifierStorageRef immutable, but then we have TransactionVerifierStorageMut to flush the changes when we're done, with a hiararchy that that can test all possible changes from one level upon the one on top of it. I'm sure there's a way to make the transaction verifier more performant by combining both, but I'm also sure we're gonna lose a dozen of security guarantees by doing that.

In short: I would prefer keeping this immutable, even if it meant we can lose some performance. Premature optimization isn't something I wanna worry about now, since the cost seems high. Usually I'm OK with it if it doesn't brick our plans.

Reply:
The motivation here is not performance but communicating correctly the semantics. In case something like signature batching is implemented, it is better to correctly communicate that something may be updated in types. That way, we know you are supposed to have a unique reference to it. If this is made immutable but mutation happens under the hood, someone in the future may take advantage of the fact that you can have multiple active references to the object. That will not result just in performance degradation but also, and more importantly, potentially *incorrect code*.

You are right that some (currently all non-test) implementations do not mutate any state and it is not ideal to expose them under a mutable ref in this case. We could change that to `&self` for now, adjusting tests to use interior mutability. However I am against keeping it as a shared ref if a mutating implementation comes along. Also, the current implementation partially mitigates the fact it's passed along mutably by having most of its members under a shared ref (e.g. the transaction is stored by shared reference), so most of the data cannot be mutated anyway.

Even with signature batching (for example), a big chunk of the context would still be immutable. It makes sense to split the context into mutable and immutable parts. I see two feasible ways of going about it:

1. Take the context as a shared ref and introduce an associated type `MyCntext::State` for the mutable part of the state.  The state part is passed along under a mutable ref, the context itself immutably. Alternatively, the state could be passed in by value and returned, to make it fell more functional, but I don't see this making much difference. This introduces more complexity in traits.
2. Leave the context ref mutable and adapt the following pattern for implementing contexts:
   ```rust
   struct MyFancyContext {
     immutable_part: &MyFancyContextRef,  // under a shared reference, does not mutate
     mutable_part: MyFancyContextMut, // owned, can freely change this
   }
   ```

Having this as a shared ref with interior mutability in presence of signature batching (or something else that updates some state in an observable way) is imho a wrong way to go about it.